File: lua/vimade/style/invert.lua
local M = {}

local CONDITION = require('vimade.style.value.condition')
local COLOR_UTIL = require('vimade.util.color')
local TYPE = require('vimade.util.type')
local VALIDATE = require('vimade.util.validate')

local INTERPOLATE_256 = COLOR_UTIL.interpolate256
local INTERPOLATE_24B = COLOR_UTIL.interpolate24b
local TO_RGB = COLOR_UTIL.toRgb

-- @required
-- number | {
--   @optional fg = 0-1,         # applies the given intensity to text
--   @optional bg = 0-1,         # applies the given intensity to background
--   @optional sp = 0-1,         # applies the given intensity to special
-- }
M.Invert = function(config)
  local result = {}
  local _value = config.value 
  local _condition = config.condition or CONDITION.INACTIVE
  result.tick = config.tick
  result.attach = function(win)
    local invert = _value
    local condition = _condition
    local style = {}
    style.win = win
    style._condition = _condition
    style._animating = false
    style.resolve = function (value, state)
      return VALIDATE.invert(TYPE.resolve_all_fn(value, style, state))
    end
    style.before = function (win, state)
      invert = style.resolve(_value, state)
      -- condition needs to be the last check due to animation checks
      if type(_condition) == 'function' then
        condition = _condition(style, state)
      end
    end
    style.key = function (win, state)
      if condition == false then
        return ''
      end
      return 'INV-' .. (invert.fg or 0) .. '-' .. (invert.bg or 0) .. '-' .. (invert.sp or 0)
    end
    style.modify = function (highlights, to_hl)
      if condition == false or invert == nil then
        return
      end
      local color
      if to_hl.fg then
        to_hl.fg = INTERPOLATE_24B(to_hl.fg, 0xFFFFFF - to_hl.fg, 1 - invert.fg)
      end
      if to_hl.bg then
        to_hl.bg = INTERPOLATE_24B(to_hl.bg, 0xFFFFFF - to_hl.bg, 1 - invert.bg)
      end
      if to_hl.sp then
        to_hl.sp = INTERPOLATE_24B(to_hl.sp, 0xFFFFFF - to_hl.sp, 1 - invert.sp)
      end
      if to_hl.ctermfg then
        color = TO_RGB(to_hl.ctermfg, true)
        to_hl.ctermfg = INTERPOLATE_256(color, {255 - color[1], 255 - color[2], 255 - color[3]}, 1 - invert.fg)
      end
      if to_hl.ctermbg then
        color = TO_RGB(to_hl.ctermbg, true)
        to_hl.ctermbg = INTERPOLATE_256(color, {255 - color[1], 255 - color[2], 255 - color[3]}, 1 - invert.bg)
      end
      for _, hl in pairs(highlights) do
        if not hl.link then
          if hl.fg then
            hl.fg = INTERPOLATE_24B(hl.fg, 0xFFFFFF - hl.fg, 1 - invert.fg)
          end
          if hl.bg then
            hl.bg = INTERPOLATE_24B(hl.bg, 0xFFFFFF - hl.bg, 1 - invert.bg)
          end
          if hl.sp then
            hl.sp = INTERPOLATE_24B(hl.sp, 0xFFFFFF - hl.sp, 1 - invert.sp)
          end
          if hl.ctermfg then
            color = TO_RGB(hl.ctermfg, true)
            hl.ctermfg = INTERPOLATE_256(color, {255 - color[1], 255 - color[2], 255 - color[3]}, 1 - invert.fg)
          end
          if hl.ctermbg then
            color = TO_RGB(hl.ctermbg, true)
            hl.ctermbg = INTERPOLATE_256(color, {255 - color[1], 255 - color[2], 255 - color[3]}, 1 - invert.bg)
          end
        end
      end
    end
    return style
  end
  result.value = function (replacement)
    if replacement ~= nil then
      _value = replacement
      return result
    end
    return _value
  end
  return result
end

return M

File: lib/vimade/style/invert.py
import sys
import time
M = sys.modules[__name__]

from vimade.util import color as COLOR_UTIL
from vimade.style.value import condition as CONDITION
from vimade.util import type as TYPE
from vimade.util import validate as VALIDATE

GLOBALS = None

def __init(args):
  global GLOBALS
  GLOBALS = args['GLOBALS']
M.__init = __init

# @required
# number | {
#   @optional 'fg': 0-1,         # applies the given intensity to text
#   @optional 'bg': 0-1,         # applies the given intensity to background
#   @optional 'sp': 0-1,         # applies the given intensity to special
# }
class Invert():
  def __init__(parent, **kwargs):
    _condition = kwargs.get('condition')
    _condition = _condition if _condition != None else CONDITION.INACTIVE
    parent._value = kwargs.get('value')
    parent.tick = kwargs.get('tick')
    class __Invert():
      def __init__(self, win, state):
        self.win = win
        self._condition = _condition
        self.condition = None
        self.invert = parent._value
        self._animating = False
      def resolve(self, value, state):
        return VALIDATE.invert(TYPE.resolve_all_fn(value, self, state))
      def before(self, win, state):
        invert = self.resolve(parent._value, state)
        self.condition = _condition(self, state) if callable(_condition) else _condition
        if self.condition == False:
          return
        self.invert = invert
      def key(self, win, state):
        if self.condition == False or not self.invert:
          return ''
        # TODO shrink
        return 'INV-' + str(self.invert.get('fg',0)) + '-' + str(self.invert.get('bg',0)) + '-' + str(self.invert.get('sp',0))
      def modify(self, hl, to_hl):
        if self.condition == False or not self.invert:
          return
        invert = self.invert
        for hi in (hl, to_hl):
          for key in ('fg', 'bg', 'sp'):
            color = hi.get(key)
            if color != None:
              hi[key] = COLOR_UTIL.interpolate24b(color, 0XFFFFFF - color, 1 - invert[key])
          for (key, i_key) in (('ctermfg', 'fg'), ('ctermbg', 'bg')):
            color = hi.get(key)
            if color != None:
              color = COLOR_UTIL.toRgb(color, True)
              target = [255 - color[0], 255 - color[1], 255 - color[2]]
              hi[key] = COLOR_UTIL.interpolate256(color, target, 1 - invert[i_key])
    parent.attach = __Invert
  def value(parent, replacement = None):
    if replacement != None:
      parent._value = replacement
      return parent
    return parent._value
